### PR TITLE
Initial commit of sagemaker-operator plugin

### PR DESCRIPTION
**Pull request by dgraeber**
### Description:

Sagemaker operator plugin - initial commit.  The helm.py was aslo modified and a new notebook added to verify plugin is working.

### Issue Reference URL

https://github.com/awslabs/aws-eks-data-maker/issues/<NUMBER>

----------
### Do not change content below. Mark applicable check boxes only.

Thank you for submitting a contribution to AWS Orbit Workbench.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a ISSUE associated with this PR?
- [ ] Does your PR title start with ISSUE-XXXX where XXXX is the issue number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [ ] Has your PR been rebased against the latest commit within the target branch (typically 'main')?
- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under Apache License 2.0?

### CODE REVIEW

File: cli/aws_orbit/remote_files/helm.py [hunk 1]
```diff
@@ -131,6 +131,14 @@ def uninstall_chart(name: str) -> None:
...
 
 
+def uninstall_chart_in_namespace(name: str, namespace: str) -> None:
+    try:
+        _logger.debug("Uninstalling %s", name)
+        sh.run(f"helm uninstall --debug --namespace {namespace} {name}")
+    except exceptions.FailedShellCommand as e:
+        _logger.error(e)
+
+
```
Comment (abaror) [inline cli/aws_orbit/remote_files/helm.py:141]: good catch , I also added a namespace to the uninstall on the kf branch

File: plugins/sm-operator/sm-operator/__init__.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+CHART_PATHS = os.path.join(os.path.dirname(__file__))
+
+@hooks.deploy
+def deploy(plugin_id: str, context: "Context", team_context: "TeamContext", parameters: Dict[str, Any]) -> None:
+    _logger.debug("Team Env name: %s | Team name: %s", context.name, team_context.name)
+    plugin_id = plugin_id.replace("_", "-")
+    _logger.debug("plugin_id: %s", plugin_id)
+    release_name = f"{team_context.name}-{plugin_id}"
+
+    ### Logic
```
Comment (abaror) [inline plugins/sm-operator/sm-operator/__init__.py:36]: I think this logic is flawed when we deal with multiple teams.  What will happen if the CRD was installed by another team.  The CRD needs to be installed once and not per team and we should separate it out of the team package.   I recommend moving the CRD into 0-commons of the env under cli/aws_orbit/data/kubectl/env/00-commons.yaml

File: plugins/sm-operator/sm-operator/__init__.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    utils.print_dir(chart_path)
+
+    repo_location = helm.init_team_repo(context=context, team_context=team_context)
+    repo = team_context.name
+    helm.add_repo(repo=repo, repo_location=repo_location)
+    chart_name, chart_version, chart_package = helm.package_chart(
+        repo=repo, chart_path=chart_path, values=vars
+    )
+   
+    if fresh_install:
```
Comment (abaror) [inline plugins/sm-operator/sm-operator/__init__.py:77]: too much.. again if we put the CRD in common, this should go back to be simple like the other plugins

File: plugins/sm-operator/sm-operator/templates/operator.yaml [hunk 4]
```diff
@@ -0,0 +1,338 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: sagemaker-k8s-operator-leader-election-role
+  namespace: {{ .Release.Namespace | default "sagemaker-k8s-operator-system" | quote }}
```
Comment (abaror) [inline plugins/sm-operator/sm-operator/templates/operator.yaml:5]: top level items must have the plugins shared labels , or future installs will have issues with upgrading. (missing helm labels)

File: plugins/sm-operator/sm-operator/templates/operator.yaml [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: sagemaker-k8s-operator-proxy-role
+subjects:
+- kind: ServiceAccount
+  name: {{ .Values.namespace }}
+  namespace: {{ .Values.namespace }}
+---
+apiVersion: v1
+kind: Service
```
Comment (abaror) [inline plugins/sm-operator/sm-operator/templates/operator.yaml:274]: same in all these places.. missing helm labels..  I am not going to add comment on every item

File: plugins/sm-operator/sm-operator/__init__.py (hunk 2 above)
Comment (abaror) [reply plugins/sm-operator/sm-operator/__init__.py:36]: Now I actually think this is bigger than this comment.  We need to create two different helm packages. One cluster level and second will be team level.  We can check if the cluster level is deployed and only update it then, and install the team level .    All the cluster level resources should go there {ClusterRole, ClusterRoleBinding(need to fix it to have unique name per team),  CRD (which I cannot find in this PR) }

Comment (abaror) [reply plugins/sm-operator/sm-operator/__init__.py:36]: I also need to understand if we are sure that we need to deploy this service once per team or once per cluster.  It looks like one per team now, so how does the notebook in each team space talks to the right controller? Lets verify this first.

File: plugins/sm-operator/sm-operator/templates/operator.yaml (hunk 4 above)
Comment (dgraeber) [reply plugins/sm-operator/sm-operator/templates/operator.yaml:5]: Show me an example and I can implement

File: plugins/sm-operator/sm-operator/templates/operator.yaml (hunk 5 above)
Comment (dgraeber) [reply plugins/sm-operator/sm-operator/templates/operator.yaml:274]: Show me an example and I can implement

File: plugins/sm-operator/sm-operator/__init__.py (hunk 2 above)
Comment (dgraeber) [reply plugins/sm-operator/sm-operator/__init__.py:36]: CRD's located here: plugins/sm-operator/sm-operator/crds/crd.yaml

I believe the service needs to be installed with a namespace to promote the isolation that Orbit promotes...?   

The notebook I provided shows an example:  current_job = !kubectl apply -f {job_yaml} -n lake-user
(though the lake user should be parameterized).

Let's have a design session?  If we put all CRD's into a separate helm package and call it when creating the env, what are the implications...?

File: cli/aws_orbit/remote_files/helm.py (hunk 1 above)
Comment (dgraeber) [reply cli/aws_orbit/remote_files/helm.py:141]: I think I should remove the other one and all calls to it....without the namespace, the uninstall fails as nothing is installed in default...

Comment (dgraeber) [reply cli/aws_orbit/remote_files/helm.py:141]: I take that back...anything installed in env will be default, so this method may be needed.  But the team plugins should call the method with the namespace...

**Review by abaror — DISMISSED** [review]

**Review by abaror — APPROVED** [review]

**Pull request merged**